File: mobile/test/permission_selection_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:kaede_mobile/src/domain/permission_selection.dart';
import 'package:kaede_mobile/src/protocol/generated.dart';

void main() {
  test('named permission selection preserves unknown high bits', () {
    final administrator = permissionMetadata.singleWhere(
      (item) => item.name == 'ADMINISTRATOR',
    );
    final futureBit = BigInt.one << 62;

    final selected = setApplicationPermission(
      futureBit,
      administrator,
      true,
    );
    expect(selected, futureBit | BigInt.from(administrator.bit));
    expect(applicationPermissionSelected(selected, administrator), isTrue);
    expect(
      setApplicationPermission(selected, administrator, false),
      futureBit,
    );
  });

  test('permission masks project readable install metadata', () {
    expect(
      selectedApplicationPermissions('3').map((item) => item.label),
      ['Create invites', 'Kick members'],
    );
    expect(
      () => applicationPermissionMask('-1'),
      throwsA(isA<FormatException>()),
    );
    expect(
      () => applicationPermissionMask('1.5'),
      throwsA(isA<FormatException>()),
    );
  });
}

File: frontend/src/lib/chat/permission-selection.test.ts
import { describe, expect, it } from 'vitest';
import {
  permissionMask,
  permissionSelected,
  selectedPermissionMetadata,
  setPermissionSelected
} from './permission-selection';

describe('named application permission selection', () => {
  it('preserves exact high and unknown bits while toggling a named permission', () => {
    const futureBit = 1n << 62n;
    const current = futureBit.toString();
    const selected = setPermissionSelected(current, 8n, true);

    expect(permissionMask(selected)).toBe(futureBit | 8n);
    expect(permissionSelected(selected, 8n)).toBe(true);
    expect(permissionMask(setPermissionSelected(selected, 8n, false))).toBe(futureBit);
  });

  it('projects readable permission labels and rejects malformed masks', () => {
    expect(selectedPermissionMetadata('3').map((item) => item.label)).toEqual([
      'Create invites',
      'Kick members'
    ]);
    expect(() => permissionMask('-1')).toThrow('non-negative whole number');
    expect(() => permissionMask('1.5')).toThrow('non-negative whole number');
  });
});
